api/envoy/service/discovery/v3/discovery.proto

```diff
@@ -179,6 +187,12 @@ message DeltaDiscoveryRequest {
   // A list of Resource names to remove from the list of tracked resources.
   repeated string resource_names_unsubscribe = 4;
 
+  // An alternative to resource_names_subscribe that specifies an optional
+  // subscription ID for each resource locator subscribed to. Resources that
+  // match each resource locator will be tagged with the associated subscription
+  // ID in the Resource.new_subscription_ids field in the response.
+  map<string, uint64> resource_locators_subscribe = 8;
```

Comment:
WDYT about using strong hashes instead of explicit resource IDs? Just trying to think about how we can simplify. The less we have in here the better.

Reply:
+1 for hashes.
At the moment it seems that the server needs to maintain for each client what subscription ID was used for a resource. With hashes, however, the server needs to store for each client the resources that it's aware of, but not "when" they were sent.

Reply:
I'm open to using a hash instead of a subscription ID assigned by the client.  But that approach adds a bit of complexity of its own:
- We'd need to either hard-code the hashing algorithm as an inherent part of the transport protocol or add a negotiation mechanism for it.
- We'd need some other way for the client to tell the server that it wants the hashes to be sent to it, maybe via a client feature.

@adisuissa The server already needs to maintain a list of all resource locators that the client is subscribed to; that's true even today.  It's true that with subscription ID, the server would also have to store the numeric subscription ID associated with each of those resource locators, but that doesn't seem like a huge increase in complexity.

Reply:
FWIW, I'm not wedded to the hashing, just wanted to make sure it was considered. I can see the value in making this client configurable. One thing to also think about (as with everything `xdstp://`), since we don't have explicit transport version negotiation, we probably want to make sure this new feature works seamlessly with existing servers.

Reply:
Right.  In this case, I think that should be fairly straightforward: the client will request the new functionality by sending the subscription IDs when it subscribes to resources.  A server that is capable of sending them will do so; a server that is not capable of sending them will just ignore them.  If a client requests the subscription IDs and the server doesn't support them, the client will be able to tell by noticing that the server did not send any subscription IDs when it sends back the first resource.

Note that we also need to require that a server that supports flexible context parameter matching must also support this subscription ID mechanism, since this mechanism is designed explicitly for the purpose of avoiding ambiguity in flexible matching.  But that seems reasonable to me, since both flexible matching and this subscription ID mechanism are in the transport protocol layer.

Reply:
In fact, I suppose we could even say that clients need to send a subscription ID only in cases where they are subscribing to a resource locator that includes a flexible context parameter matcher.  That way, we're not adding any wire overhead for existing cases, which don't really need this mechanism.